File: src/ralph_wiggum/core/loop_runner.py
"""Main loop orchestration logic."""

import sys
from pathlib import Path

from rich.console import Console

from ralph_wiggum.core.kiro_client import KiroClient
from ralph_wiggum.core.session_reader import get_latest_session
from ralph_wiggum.models.config import LoopConfig
from ralph_wiggum.models.state import LoopState

STATE_FILE = Path(".kiro/ralph-loop.local.md")


def run_loop(config: LoopConfig, console: Console) -> None:
    """Run the Ralph loop.

    Parameters
    ----------
    config
        Loop configuration with prompt and settings.
    console
        Rich console for output.
    """
    client = KiroClient(agent_name=config.agent_name)
    cwd = Path.cwd()

    console.print("[bold blue]Ralph loop starting[/bold blue]")
    console.print(f"   Min iterations: {config.min_iterations}")
    console.print(f"   Max iterations: {config.max_iterations or 'unlimited'}")
    console.print(f"   Completion promise: {config.completion_promise}")
    console.print()

    iteration = 0
    try:
        while True:
            iteration += 1

            # Create/update state file
            state = LoopState(
                iteration=iteration,
                min_iterations=config.min_iterations,
                max_iterations=config.max_iterations,
                completion_promise=config.completion_promise,
                prompt=config.prompt,
            )
            STATE_FILE.parent.mkdir(parents=True, exist_ok=True)
            STATE_FILE.write_text(state.to_markdown())

            console.print(f"[yellow]Iteration {iteration}[/yellow]")

            # Run kiro-cli
            exit_code = client.run_chat(config.prompt)

            if exit_code != 0:
                console.print(f"[red]Kiro exited with code {exit_code}[/red]")

            # Only check for completion after minimum iterations reached
            if iteration >= config.min_iterations:
                session = get_latest_session(cwd)
                if session and session.check_completion_promise(
                    config.completion_promise
                ):
                    console.print(f"[green]Completed at iteration {iteration}![/green]")
                    cleanup()
                    sys.exit(0)
            else:
                console.print(
                    f"[dim]Iteration {iteration}/{config.min_iterations} "
                    f"(min not reached, skipping completion check)[/dim]"
                )

            # Check max iterations
            if config.max_iterations > 0 and iteration >= config.max_iterations:
                console.print(
                    f"[red]Max iterations ({config.max_iterations}) reached[/red]"
                )
                cleanup()
                sys.exit(0)

    except KeyboardInterrupt:
        console.print(f"\n[yellow]Interrupted at iteration {iteration}[/yellow]")
        cleanup()
        sys.exit(1)


def cleanup() -> None:
    """Remove state files."""
    STATE_FILE.unlink(missing_ok=True)

File: src/core/loop-runner.ts
/**
 * @fileoverview Main loop orchestration for Ralph Wiggum.
 * Manages the iterative loop that runs kiro-cli until completion.
 * @module core/loop-runner
 */
import { mkdir, unlink } from "node:fs/promises";
import { dirname } from "node:path";
import { log } from "@clack/prompts";
import pc from "picocolors";

import type { LoopConfig } from "../schemas/config";
import {
	checkCompletionPromise,
	extractRalphFeedback,
	type RalphFeedback,
} from "../schemas/session";
import { type LoopState, stateToJson } from "../schemas/state";
import { STATE_FILE } from "../utils/paths";
import { KiroClient } from "./kiro-client";
import { getLatestSession } from "./session-reader";

/**
 * Runs the Ralph Wiggum iterative loop.
 * Executes kiro-cli repeatedly until the completion promise is detected
 * or maximum iterations are reached.
 * @param config - Loop configuration with prompt and iteration settings
 * @returns Resolves when the loop completes, is interrupted, or reaches max iterations
 * @example
 * ```typescript
 * await runLoop({
 *   prompt: "Build a REST API",
 *   minIterations: 1,
 *   maxIterations: 10,
 *   completionPromise: "DONE",
 *   agentName: null
 * });
 * ```
 */
export async function runLoop(config: LoopConfig): Promise<void> {
	const client = new KiroClient(config.agentName);
	const cwd = process.cwd();

	// Cleanup function to mark state as inactive (preserves for resume)
	const cleanup = async (
		currentIteration?: number,
		currentFeedback?: RalphFeedback,
	): Promise<void> => {
		// If we have iteration info, preserve the state for resume
		if (currentIteration !== undefined && currentIteration > 0) {
			try {
				const state: LoopState = {
					active: false, // Mark as inactive
					iteration: currentIteration,
					minIterations: config.minIterations,
					maxIterations: config.maxIterations,
					completionPromise: config.completionPromise,
					startedAt: new Date().toISOString(),
					prompt: config.prompt,
					previousFeedback: currentFeedback,
				};
				await mkdir(dirname(STATE_FILE), { recursive: true });
				await Bun.write(STATE_FILE, stateToJson(state));
			} catch {
				// If we can't write state, just delete it
				await unlink(STATE_FILE).catch(() => {});
			}
		} else {
			// No iteration info, just delete the state file
			await unlink(STATE_FILE).catch(() => {});
		}
	};

	// Clean any stale state file from a previous crashed loop
	// This prevents "already complete" issues when starting a new loop
	// Skip cleanup if this is a resume operation
	if (!config.isResume) {
		await cleanup();
	}

	// Display startup info
	const loopType = config.isResume
		? "Ralph loop resuming"
		: "Ralph loop starting";
	log.info(pc.bold(pc.blue(loopType)));
	log.message(`   Min iterations: ${config.minIterations}`);
	log.message(`   Max iterations: ${config.maxIterations || "unlimited"}`);
	log.message(`   Completion promise: ${config.completionPromise}`);
	if (config.isResume) {
		log.message(`   Resuming from iteration: ${config.resumeFromIteration}`);
	}
	console.log();

	let iteration = config.isResume ? config.resumeFromIteration : 0;
	let previousFeedback: RalphFeedback | undefined;

	// Handle Ctrl+C gracefully
	process.on("SIGINT", async () => {
		log.warn(pc.yellow(`\nInterrupted at iteration ${iteration}`));
		log.message("Saving state for resume...");
		await cleanup(iteration, previousFeedback);
		log.message("Run 'ralph resume' to continue where you left off.");
		process.exit(1);
	});

	try {
		while (true) {
			iteration++;

			// Create/update state file
			const state: LoopState = {
				active: true,
				iteration,
				minIterations: config.minIterations,
				maxIterations: config.maxIterations,
				completionPromise: config.completionPromise,
				startedAt: new Date().toISOString(),
				prompt: config.prompt,
				previousFeedback,
			};

			// Ensure directory exists and write state file
			await mkdir(dirname(STATE_FILE), { recursive: true });
			await Bun.write(STATE_FILE, stateToJson(state));

			// Log iteration start
			log.step(pc.yellow(`Iteration ${iteration}`));

			// Run kiro-cli
			const exitCode = await client.runChat(config.prompt);

			if (exitCode !== 0) {
				log.warn(pc.red(`Kiro exited with code ${exitCode}`));
			}

			// Small delay to ensure Kiro has fully released the database
			// and any WAL writes have synced
			await new Promise((resolve) => setTimeout(resolve, 500));

			// Get the session to check completion and extract feedback
			// Wrap in try-catch to handle any database/memory issues gracefully
			let session = null;
			try {
				session = getLatestSession(cwd);
			} catch (err) {
				log.warn(pc.dim(`Could not read session: ${err}`));
			}

			// Extract feedback from this iteration for the next
			// Only if session looks valid (has history array)
			if (session?.history && Array.isArray(session.history)) {
				try {
					previousFeedback = extractRalphFeedback(session) ?? undefined;
					if (previousFeedback?.qualityScore !== undefined) {
						log.message(
							pc.dim(`Quality score: ${previousFeedback.qualityScore}/10`),
						);
					}
				} catch (err) {
					log.warn(pc.dim(`Could not extract feedback: ${err}`));
					previousFeedback = undefined;
				}
			}

			// Only check for completion after minimum iterations reached
			if (iteration >= config.minIterations) {
				if (
					session &&
					checkCompletionPromise(session, config.completionPromise)
				) {
					log.success(pc.green(`Completed at iteration ${iteration}!`));
					// Task completed successfully - can delete state
					await cleanup();
					process.exit(0);
				}
			} else {
				log.message(
					pc.dim(
						`Iteration ${iteration}/${config.minIterations} (min not reached, skipping completion check)`,
					),
				);
			}

			// Check max iterations
			if (config.maxIterations > 0 && iteration >= config.maxIterations) {
				log.warn(pc.red(`Max iterations (${config.maxIterations}) reached`));
				log.message("Saving state for resume...");
				await cleanup(iteration, previousFeedback);
				log.message("Run 'ralph resume' to continue where you left off.");
				process.exit(0);
			}
		}
	} catch (error) {
		// Re-throw non-interrupt errors
		if (error instanceof Error && !error.message.includes("SIGINT")) {
			throw error;
		}
	}
}
